File: app/src/main/java/com/example/task2/javafolder/viewviewmodel/fragments/parentfragment/ParentFragment.java
package com.example.task2.javafolder.viewviewmodel.fragments.parentfragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;

import com.example.task2.databinding.FragmentWithBenchmarksBinding;
import com.example.task2.javafolder.model.StateOfCalculation;
import com.example.task2.javafolder.viewviewmodel.fragments.BaseFragment;
import com.example.task2.javafolder.viewviewmodel.fragments.BenchmarksAdapter;
import com.google.android.material.textfield.TextInputEditText;

import java.util.ArrayList;

public abstract class ParentFragment extends BaseFragment {

    protected FragmentWithBenchmarksBinding binding;

    private BenchmarksAdapter benchmarksAdapter;

    public abstract int getSpanCount();

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentWithBenchmarksBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setUpView();
        setupControls();
        subscribeToViewModel();
    }

    @Override
    public abstract ParentViewModel determineViewModelProvider();



    @Override
    public TextInputEditText getUsersInputInstance() {
        return binding.usersInput;
    }

    @Override
    public Button getButtonInstance() {
        return binding.startStopButton;
    }

    private void setUpView() {
        benchmarksAdapter = new BenchmarksAdapter();
        setupRecyclerViewElements();
        determineViewModelProvider().setBenchmarks();
        setNumberToEditText();
    }

    private void setupControls() {
        binding.startStopButton.setOnClickListener(view -> {
            if (binding.startStopButton.isChecked()) {
                determineViewModelProvider().startCalculations();
            } else {
                determineViewModelProvider().stopCalculations();
            }
        });
    }

    private void subscribeToViewModel() {
        determineViewModelProvider().liveDataForStateOfCalculation.observe(getViewLifecycleOwner(), stateOfCalculation ->
                binding.startStopButton.setChecked(stateOfCalculation != StateOfCalculation.TERMINATED));
        determineViewModelProvider().liveDataForListOfBenchmarks.observe(getViewLifecycleOwner(), list ->
                benchmarksAdapter.submitList(new ArrayList<>(list)));
    }

    private void setupRecyclerViewElements() {
        binding.recyclerViewForBenchmarks.setLayoutManager
                (new GridLayoutManager(getContext(), getSpanCount()));
        binding.recyclerViewForBenchmarks.setHasFixedSize(true);
        binding.recyclerViewForBenchmarks.setAdapter(benchmarksAdapter);
    }

    private void setNumberToEditText() {
        binding.usersInput.setText(determineViewModelProvider().getUsersInputForTextInput());
    }
}

File: app/src/main/java/com/example/task2/kotlinfolder/viewviewmodel/fragments/parentfragment/ParentFragment.kt
package com.example.task2.kotlinfolder.viewviewmodel.fragments.parentfragment

import android.widget.Button
import androidx.recyclerview.widget.GridLayoutManager
import com.example.task2.databinding.FragmentWithBenchmarksBinding
import com.example.task2.kotlinfolder.model.StateOfCalculation
import com.example.task2.kotlinfolder.viewviewmodel.fragments.BaseFragment
import com.example.task2.kotlinfolder.viewviewmodel.fragments.BenchmarkAdapter
import com.google.android.material.textfield.TextInputEditText

abstract class ParentFragment<VM : ParentViewModel>(private val benchmarkAdapter: BenchmarkAdapter) :
    BaseFragment<FragmentWithBenchmarksBinding, VM>() {

    override fun getViewBinding() = FragmentWithBenchmarksBinding.inflate(layoutInflater)

    override fun initViews() {
        super.initViews()
        setUpView()
        setUpControls()
        subscribeToViewModel()
    }

    abstract val spanCount: Int

    override val usersInputEditText: TextInputEditText get() = binding.usersInput

    override val calculateButton: Button get() = binding.startStopButton


    private fun setUpView() {
        setupRecyclerViewElements()
        setNumberToEditText()
    }

    private fun setUpControls() {
        binding.startStopButton.setOnClickListener {
            when (binding.startStopButton.isChecked) {
                true -> viewModel.startCalculations()
                false -> viewModel.stopCalculations()
            }
        }
    }

    private fun subscribeToViewModel() {
        viewModel.liveDataForStateOfCalculation.observe(viewLifecycleOwner) {
            binding.startStopButton.isChecked = it != StateOfCalculation.TERMINATED
        }
        viewModel.liveDataForListOfBenchmarks.observe(viewLifecycleOwner) {
            this.benchmarkAdapter.submitList(ArrayList(it))
        }
    }

    private fun setupRecyclerViewElements() {
        binding.recyclerViewForBenchmarks.layoutManager = GridLayoutManager(context, spanCount)
        binding.recyclerViewForBenchmarks.setHasFixedSize(true)
        binding.recyclerViewForBenchmarks.adapter = this.benchmarkAdapter
    }

    private fun setNumberToEditText() {
        binding.usersInput.setText(viewModel.getUsersInputForTextInput())
    }

}